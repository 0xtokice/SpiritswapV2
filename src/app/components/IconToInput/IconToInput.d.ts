import { ReactNode, ChangeEvent } from 'react';
import { InputProps } from '../Input';

export interface StyleProps {
  open?: boolean;
  iconPos?: 'left' | 'right';
  disabled?: boolean;
}

export interface Props extends InputProps, StyleProps {
  icon?: ReactNode;
  placeholder?: string;
  onChange?: (value: ChangeEvent<HTMLInputElement>) => void;
}
