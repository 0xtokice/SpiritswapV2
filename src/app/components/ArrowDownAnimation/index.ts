export { default as ArrowDownAnimation } from './ArrowDown';
