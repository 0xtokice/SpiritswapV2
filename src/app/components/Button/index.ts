export type { Props as ButtonProps } from './Button.d';
export { default as Button } from './Button';
