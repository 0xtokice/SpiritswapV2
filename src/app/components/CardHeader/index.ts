export type { Props as CardHeaderProps } from './CardHeader.d';
export { default as CardHeader } from './CardHeader';
