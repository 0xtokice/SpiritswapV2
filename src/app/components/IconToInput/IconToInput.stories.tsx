import { ComponentMeta, ComponentStory } from '@storybook/react';
import IconToInput from './IconToInput';
import { Props } from './IconToInput.d';
import { ChangeEvent, useCallback, useState } from 'react';

export default {
  title: 'SpiritSwap/Components/IconToInput',
  component: IconToInput,
  args: {
    disabled: false,
    open: false,
    placeholder: `Search for token...`,
  },
  argTypes: {
    onChange: { action: 'onChange' },
  },
} as ComponentMeta<typeof IconToInput>;

const Template: ComponentStory<typeof IconToInput> = (args: Props) => (
  <IconToInput {...args} />
);

const ControlledTemplate: ComponentStory<typeof IconToInput> = args => {
  const [controlledValue, setControlledValue] = useState(
    'Try to type something...',
  );
  const onChangeCallback = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      setControlledValue(event.target.value.replace(/test/g, 'hurray'));
    },
    [setControlledValue],
  );
  return (
    <IconToInput
      {...args}
      value={controlledValue}
      onChange={onChangeCallback}
    />
  );
};

export const Default = Template.bind({});
Default.args = {};

export const Open = Template.bind({});
Open.args = {
  open: true,
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const OpenAndDisabled = Template.bind({});
OpenAndDisabled.args = {
  disabled: true,
  open: true,
};

export const Controlled = ControlledTemplate.bind({});
Controlled.args = {};
