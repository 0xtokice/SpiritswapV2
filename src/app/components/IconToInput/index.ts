export type { Props as IconToInputProps } from './IconToInput.d';
export { default as IconToInput } from './IconToInput';
