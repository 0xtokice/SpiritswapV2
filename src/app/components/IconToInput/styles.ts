import styled from 'styled-components';
import { animated, disabled as disabledStyles } from 'app/components/shared';
import type { StyleProps } from './IconToInput.d';

export const StyledWrapper = styled.div<StyleProps>`
  display: flex;
  flex-direction: row;
  align-items: center;
  border-radius: ${({ theme }) => `${theme.borderRadius.md}`};
  flex-direction: ${({ iconPos }) =>
    iconPos === 'right' ? 'row-reverse' : 'row'};

  &,
  & > div {
    width: 100%;
  }

  ${animated}

  ${({ disabled, open }) => disabled && !open && disabledStyles}
`;
