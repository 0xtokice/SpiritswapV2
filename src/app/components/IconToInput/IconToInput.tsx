import { useState, useEffect, useCallback } from 'react';
import { ReactComponent as CloseIcon } from 'app/assets/images/times.svg';
import { ReactComponent as SearchIcon } from 'app/assets/images/search-loupe.svg';
import Input from '../Input/Input';
import IconButton from '../IconButton/IconButton';
import type { Props } from './IconToInput.d';
import { StyledWrapper } from './styles';

const IconToInput = ({
  icon = <SearchIcon />,
  disabled = false,
  iconPos,
  open = false,
  className,
  ...props
}: Props) => {
  const [isOpen, setIsOpen] = useState<boolean>(open);

  const onIconClick = useCallback(() => {
    if (!isOpen && !disabled) {
      setIsOpen(true);
    }
    // TODO Implmement React.forwardRef() in Input
    // inputRef.current?.focus(), 0);
  }, [isOpen, disabled]);

  const onCloseButtonClick = useCallback(() => {
    if (!disabled) {
      setIsOpen(false);
    }
  }, [disabled]);

  useEffect(() => {
    setIsOpen(isOpen);
  }, [isOpen]);

  useEffect(() => {
    setIsOpen(open);
  }, [open]);

  return (
    <StyledWrapper
      className={className}
      iconPos={iconPos}
      open={isOpen}
      disabled={disabled}
    >
      {!isOpen ? (
        <IconButton
          flat={true}
          icon={icon}
          onClick={onIconClick}
          variant={'secondary'}
          size={'big'}
        />
      ) : (
        <Input
          disabled={disabled}
          iconPrefix={icon}
          iconSuffix={<CloseIcon />}
          onClickIconSuffix={onCloseButtonClick}
          {...props}
        />
      )}
    </StyledWrapper>
  );
};

export default IconToInput;
