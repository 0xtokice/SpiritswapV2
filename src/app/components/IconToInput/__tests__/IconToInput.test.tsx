import { render, screen, fireEvent } from '@testing-library/react';
import { IconToInput, IconToInputProps } from 'app/components/IconToInput';
import { ThemeWrapper } from 'app/components/shared/testing/ThemeWrapper';
import { ReactComponent as LockIcon } from 'app/assets/images/lock.svg';

const Component = (props: IconToInputProps) => {
  return (
    <ThemeWrapper>
      <IconToInput {...props} />
    </ThemeWrapper>
  );
};

describe('IconToInput component', () => {
  test('renders without crashing', () => {
    const { container } = render(<Component />);
    expect(container.firstChild).toBeTruthy();
  });

  test('opened', () => {
    const { container } = render(<Component open={true} />);
    const element = container.querySelector('input');
    expect(element).toBeTruthy();
  });

  test('closed', () => {
    const { container } = render(<Component open={false} />);
    const element = container.querySelector('input');
    expect(element).toBeFalsy();
  });

  test('open and disabled', () => {
    const { container } = render(<Component open={true} disabled={true} />);
    const element = container.querySelector('input');
    expect(element).toHaveAttribute('disabled');
  });

  test('with custom icon', () => {
    const icon = <LockIcon />;
    const { container } = render(<Component open={true} icon={icon} />);
    const element = container.querySelector('svg');
    expect(element).toMatchSnapshot('<svg>lock.svg</svg>');
  });

  test('placeholder', () => {
    const placeholder: string = 'Search for token...';
    render(<Component open={true} placeholder={placeholder} />);
    screen.getByPlaceholderText(placeholder);
  });

  test('onChange', () => {
    const demoInput: string = 'Demo input';
    const { container } = render(<Component open={true} />);
    const input = container.querySelector('input');
    expect(input).toBeTruthy();

    if (input) {
      fireEvent.change(input, { target: { value: demoInput } });
      expect(input.value).toBe(demoInput);
    }
  });
});
