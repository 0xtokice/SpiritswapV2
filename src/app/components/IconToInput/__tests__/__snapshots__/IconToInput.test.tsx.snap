// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconToInput component with custom icon: <svg>lock.svg</svg> 1`] = `
<svg>
  lock.svg
</svg>
`;
